File: .UNSAFE-ALL-BINDINGS-DELETED-FINAL-20260125/bindings/clojure/src/proven/safe_phone.clj
;; SPDX-License-Identifier: PMPL-1.0
;; SPDX-FileCopyrightText: 2025 Hyperpolymath

(ns proven.safe-phone
  "Safe phone number validation following E.164 standard."
  (:require [clojure.string :as str]))

;; Country calling codes (ITU-T E.164)
(def country-codes
  "Map of country code values to country identifiers."
  {1   :US    ; USA, Canada, Caribbean
   7   :RU    ; Russia, Kazakhstan
   20  :EG    ; Egypt
   27  :ZA    ; South Africa
   30  :GR    ; Greece
   31  :NL    ; Netherlands
   32  :BE    ; Belgium
   33  :FR    ; France
   34  :ES    ; Spain
   36  :HU    ; Hungary
   39  :IT    ; Italy
   40  :RO    ; Romania
   41  :CH    ; Switzerland
   43  :AT    ; Austria
   44  :UK    ; United Kingdom
   45  :DK    ; Denmark
   46  :SE    ; Sweden
   47  :NO    ; Norway
   48  :PL    ; Poland
   49  :DE    ; Germany
   51  :PE    ; Peru
   52  :MX    ; Mexico
   53  :CU    ; Cuba
   54  :AR    ; Argentina
   55  :BR    ; Brazil
   56  :CL    ; Chile
   57  :CO    ; Colombia
   58  :VE    ; Venezuela
   60  :MY    ; Malaysia
   61  :AU    ; Australia
   62  :ID    ; Indonesia
   63  :PH    ; Philippines
   64  :NZ    ; New Zealand
   65  :SG    ; Singapore
   66  :TH    ; Thailand
   81  :JP    ; Japan
   82  :KR    ; South Korea
   84  :VN    ; Vietnam
   86  :CN    ; China
   90  :TR    ; Turkey
   91  :IN    ; India
   92  :PK    ; Pakistan
   93  :AF    ; Afghanistan
   94  :LK    ; Sri Lanka
   95  :MM    ; Myanmar
   98  :IR    ; Iran
   212 :MA    ; Morocco
   213 :DZ    ; Algeria
   216 :TN    ; Tunisia
   218 :LY    ; Libya
   220 :GM    ; Gambia
   221 :SN    ; Senegal
   234 :NG    ; Nigeria
   254 :KE    ; Kenya
   255 :TZ    ; Tanzania
   256 :UG    ; Uganda
   260 :ZM    ; Zambia
   263 :ZW    ; Zimbabwe
   351 :PT    ; Portugal
   352 :LU    ; Luxembourg
   353 :IE    ; Ireland
   354 :IS    ; Iceland
   355 :AL    ; Albania
   356 :MT    ; Malta
   357 :CY    ; Cyprus
   358 :FI    ; Finland
   359 :BG    ; Bulgaria
   370 :LT    ; Lithuania
   371 :LV    ; Latvia
   372 :EE    ; Estonia
   380 :UA    ; Ukraine
   381 :RS    ; Serbia
   385 :HR    ; Croatia
   386 :SI    ; Slovenia
   420 :CZ    ; Czech Republic
   421 :SK    ; Slovakia
   852 :HK    ; Hong Kong
   853 :MO    ; Macau
   855 :KH    ; Cambodia
   856 :LA    ; Laos
   880 :BD    ; Bangladesh
   886 :TW    ; Taiwan
   960 :MV    ; Maldives
   961 :LB    ; Lebanon
   962 :JO    ; Jordan
   963 :SY    ; Syria
   964 :IQ    ; Iraq
   965 :KW    ; Kuwait
   966 :SA    ; Saudi Arabia
   967 :YE    ; Yemen
   968 :OM    ; Oman
   970 :PS    ; Palestine
   971 :AE    ; UAE
   972 :IL    ; Israel
   973 :BH    ; Bahrain
   974 :QA    ; Qatar
   975 :BT    ; Bhutan
   976 :MN    ; Mongolia
   977 :NP    ; Nepal
   992 :TJ    ; Tajikistan
   993 :TM    ; Turkmenistan
   994 :AZ    ; Azerbaijan
   995 :GE    ; Georgia
   996 :KG    ; Kyrgyzstan
   998 :UZ    ; Uzbekistan})

(def country-code-lengths
  "Set of valid country code lengths for parsing."
  #{1 2 3})

(defrecord PhoneNumber [country-code national-number raw-input])

(defn- extract-digits
  "Extract only digit characters from a string."
  [phone-string]
  (apply str (filter #(Character/isDigit ^char %) phone-string)))

(defn- parse-country-code
  "Try to parse country code from digits. Returns [country-code remaining-digits] or nil."
  [digits]
  (loop [code-length 3]
    (when (>= code-length 1)
      (if (>= (count digits) code-length)
        (let [potential-code (Long/parseLong (subs digits 0 code-length))]
          (if (contains? country-codes potential-code)
            [potential-code (subs digits code-length)]
            (recur (dec code-length))))
        (recur (dec code-length))))))

(defn parse
  "Parse a phone number string.
   Returns {:ok PhoneNumber} or {:error message}."
  [phone-string]
  (cond
    (str/blank? phone-string)
    {:error "empty_phone_number"}

    :else
    (let [trimmed-input (str/trim phone-string)
          digits (extract-digits trimmed-input)]
      (cond
        (< (count digits) 7)
        {:error "phone_number_too_short"}

        (> (count digits) 15)
        {:error "phone_number_too_long"}

        :else
        (if-let [[country-code-value national-digits] (parse-country-code digits)]
          (if (< (count national-digits) 4)
            {:error "national_number_too_short"}
            {:ok (->PhoneNumber country-code-value national-digits trimmed-input)})
          {:error "unknown_country_code"})))))

(defn parse!
  "Parse a phone number string. Throws ex-info on error."
  [phone-string]
  (let [result (parse phone-string)]
    (if (:ok result)
      (:ok result)
      (throw (ex-info (str "Invalid phone number: " (:error result))
                      {:type :invalid-phone-number
                       :error (:error result)
                       :input phone-string})))))

(defn valid?
  "Check if a string is a valid phone number."
  [phone-string]
  (contains? (parse phone-string) :ok))

(defn get-country-code
  "Get the numeric country code."
  [phone-number]
  (:country-code phone-number))

(defn get-country
  "Get the country identifier keyword."
  [phone-number]
  (get country-codes (:country-code phone-number) :unknown))

(defn get-national-number
  "Get the national number (without country code)."
  [phone-number]
  (:national-number phone-number))

(defn format-e164
  "Format phone number in E.164 format (+CC...)."
  [phone-number]
  (str "+" (:country-code phone-number) (:national-number phone-number)))

(defn format-international
  "Format phone number with spaces for readability."
  [phone-number]
  (let [country-code-str (str (:country-code phone-number))
        national-number (:national-number phone-number)
        national-length (count national-number)]
    (cond
      (<= national-length 4)
      (str "+" country-code-str " " national-number)

      (<= national-length 7)
      (str "+" country-code-str " "
           (subs national-number 0 3) " "
           (subs national-number 3))

      (<= national-length 10)
      (str "+" country-code-str " "
           (subs national-number 0 3) " "
           (subs national-number 3 6) " "
           (subs national-number 6))

      :else
      (str "+" country-code-str " " national-number))))

(defn format-national
  "Format as national number (without country code)."
  [phone-number]
  (let [national-number (:national-number phone-number)
        national-length (count national-number)]
    (cond
      (<= national-length 4)
      national-number

      (<= national-length 7)
      (str (subs national-number 0 3) "-"
           (subs national-number 3))

      (<= national-length 10)
      (str "(" (subs national-number 0 3) ") "
           (subs national-number 3 6) "-"
           (subs national-number 6))

      :else
      national-number)))

(defn format-rfc3966
  "Format phone number as RFC 3966 tel: URI."
  [phone-number]
  (str "tel:+" (:country-code phone-number) (:national-number phone-number)))

(defn digit-count
  "Get total digit count (country code + national number)."
  [phone-number]
  (let [country-code-digits (count (str (:country-code phone-number)))]
    (+ country-code-digits (count (:national-number phone-number)))))

(defn from-country?
  "Check if phone number is from a specific country."
  [phone-number country-keyword]
  (= (get-country phone-number) country-keyword))

(defn from-countries?
  "Check if phone number is from one of the specified countries."
  [phone-number country-keywords]
  (contains? (set country-keywords) (get-country phone-number)))

(defn normalize
  "Normalize a phone number to E.164 format string."
  [phone-string]
  (when-let [parsed (:ok (parse phone-string))]
    (format-e164 parsed)))

(defn equivalent?
  "Check if two phone numbers are equivalent (same E.164)."
  [phone-a phone-b]
  (let [normalized-a (normalize phone-a)
        normalized-b (normalize phone-b)]
    (and normalized-a normalized-b (= normalized-a normalized-b))))

(defn obfuscate
  "Obfuscate phone number for display (e.g., +1 ***-***-4567)."
  [phone-number]
  (let [national-number (:national-number phone-number)
        national-length (count national-number)]
    (if (<= national-length 4)
      (str "+" (:country-code phone-number) " " national-number)
      (let [visible-suffix (subs national-number (- national-length 4))
            masked-prefix (apply str (repeat (- national-length 4) "*"))]
        (str "+" (:country-code phone-number) " " masked-prefix visible-suffix)))))

File: bindings/kotlin/src/main/kotlin/com/hyperpolymath/proven/SafePhone.kt
// SPDX-License-Identifier: PMPL-1.0-or-later
// SPDX-FileCopyrightText: 2025 Hyperpolymath

package com.hyperpolymath.proven

/**
 * Phone number type classification.
 */
enum class PhoneNumberType {
    MOBILE,
    FIXED_LINE,
    TOLL_FREE,
    PREMIUM_RATE,
    SHARED_COST,
    VOIP,
    PERSONAL_NUMBER,
    PAGER,
    UAN,  // Universal Access Number
    UNKNOWN
}

/**
 * ITU country calling codes (common ones).
 *
 * @property code The numeric country calling code value.
 * @property displayPrefix The formatted prefix with + sign.
 */
enum class CountryCode(val code: Int, val displayPrefix: String) {
    CC1(1, "+1"),         // USA, Canada, Caribbean
    CC7(7, "+7"),         // Russia, Kazakhstan
    CC20(20, "+20"),      // Egypt
    CC27(27, "+27"),      // South Africa
    CC30(30, "+30"),      // Greece
    CC31(31, "+31"),      // Netherlands
    CC32(32, "+32"),      // Belgium
    CC33(33, "+33"),      // France
    CC34(34, "+34"),      // Spain
    CC36(36, "+36"),      // Hungary
    CC39(39, "+39"),      // Italy
    CC40(40, "+40"),      // Romania
    CC41(41, "+41"),      // Switzerland
    CC43(43, "+43"),      // Austria
    CC44(44, "+44"),      // UK
    CC45(45, "+45"),      // Denmark
    CC46(46, "+46"),      // Sweden
    CC47(47, "+47"),      // Norway
    CC48(48, "+48"),      // Poland
    CC49(49, "+49"),      // Germany
    CC51(51, "+51"),      // Peru
    CC52(52, "+52"),      // Mexico
    CC53(53, "+53"),      // Cuba
    CC54(54, "+54"),      // Argentina
    CC55(55, "+55"),      // Brazil
    CC56(56, "+56"),      // Chile
    CC57(57, "+57"),      // Colombia
    CC58(58, "+58"),      // Venezuela
    CC60(60, "+60"),      // Malaysia
    CC61(61, "+61"),      // Australia
    CC62(62, "+62"),      // Indonesia
    CC63(63, "+63"),      // Philippines
    CC64(64, "+64"),      // New Zealand
    CC65(65, "+65"),      // Singapore
    CC66(66, "+66"),      // Thailand
    CC81(81, "+81"),      // Japan
    CC82(82, "+82"),      // South Korea
    CC84(84, "+84"),      // Vietnam
    CC86(86, "+86"),      // China
    CC90(90, "+90"),      // Turkey
    CC91(91, "+91"),      // India
    CC92(92, "+92"),      // Pakistan
    CC93(93, "+93"),      // Afghanistan
    CC94(94, "+94"),      // Sri Lanka
    CC95(95, "+95"),      // Myanmar
    CC98(98, "+98"),      // Iran
    CC212(212, "+212"),   // Morocco
    CC213(213, "+213"),   // Algeria
    CC216(216, "+216"),   // Tunisia
    CC218(218, "+218"),   // Libya
    CC234(234, "+234"),   // Nigeria
    CC254(254, "+254"),   // Kenya
    CC255(255, "+255"),   // Tanzania
    CC256(256, "+256"),   // Uganda
    CC351(351, "+351"),   // Portugal
    CC352(352, "+352"),   // Luxembourg
    CC353(353, "+353"),   // Ireland
    CC354(354, "+354"),   // Iceland
    CC358(358, "+358"),   // Finland
    CC380(380, "+380"),   // Ukraine
    CC381(381, "+381"),   // Serbia
    CC385(385, "+385"),   // Croatia
    CC386(386, "+386"),   // Slovenia
    CC420(420, "+420"),   // Czech Republic
    CC421(421, "+421"),   // Slovakia
    CC852(852, "+852"),   // Hong Kong
    CC853(853, "+853"),   // Macau
    CC855(855, "+855"),   // Cambodia
    CC856(856, "+856"),   // Laos
    CC880(880, "+880"),   // Bangladesh
    CC886(886, "+886"),   // Taiwan
    CC960(960, "+960"),   // Maldives
    CC961(961, "+961"),   // Lebanon
    CC962(962, "+962"),   // Jordan
    CC963(963, "+963"),   // Syria
    CC964(964, "+964"),   // Iraq
    CC965(965, "+965"),   // Kuwait
    CC966(966, "+966"),   // Saudi Arabia
    CC967(967, "+967"),   // Yemen
    CC968(968, "+968"),   // Oman
    CC971(971, "+971"),   // UAE
    CC972(972, "+972"),   // Israel
    CC973(973, "+973"),   // Bahrain
    CC974(974, "+974"),   // Qatar
    CC975(975, "+975"),   // Bhutan
    CC976(976, "+976"),   // Mongolia
    CC977(977, "+977"),   // Nepal
    CC992(992, "+992"),   // Tajikistan
    CC993(993, "+993"),   // Turkmenistan
    CC994(994, "+994"),   // Azerbaijan
    CC995(995, "+995"),   // Georgia
    CC996(996, "+996"),   // Kyrgyzstan
    CC998(998, "+998");   // Uzbekistan

    companion object {
        private val byCode: Map<Int, CountryCode> = entries.associateBy { it.code }

        /**
         * Find country code by numeric value.
         */
        fun fromCode(code: Int): CountryCode? = byCode[code]

        /**
         * Parse country code from string digits.
         */
        fun parse(digits: String): Pair<CountryCode, String>? {
            // Try 3-digit codes first, then 2-digit, then 1-digit
            for (length in 3 downTo 1) {
                if (digits.length >= length) {
                    val prefix = digits.substring(0, length)
                    val codeValue = prefix.toIntOrNull()
                    if (codeValue != null) {
                        val countryCode = fromCode(codeValue)
                        if (countryCode != null) {
                            return Pair(countryCode, digits.substring(length))
                        }
                    }
                }
            }
            return null
        }
    }
}

/**
 * Phone number parsing errors.
 */
sealed class PhoneError {
    data class InvalidCharacter(val character: Char, val position: Int) : PhoneError()
    data class TooShort(val length: Int) : PhoneError()
    data class TooLong(val length: Int) : PhoneError()
    data class InvalidCountryCode(val code: String) : PhoneError()
    data class InvalidNationalNumber(val number: String) : PhoneError()
    data object EmptyInput : PhoneError()

    override fun toString(): String = when (this) {
        is InvalidCharacter -> "Invalid character '$character' at position $position"
        is TooShort -> "Phone number too short: $length digits"
        is TooLong -> "Phone number too long: $length digits"
        is InvalidCountryCode -> "Invalid country code: $code"
        is InvalidNationalNumber -> "Invalid national number: $number"
        is EmptyInput -> "Empty input"
    }
}

/**
 * Validated phone number in E.164 format.
 *
 * @property countryCode The country calling code.
 * @property nationalNumber The national number (without country code).
 */
data class PhoneNumber(
    val countryCode: CountryCode,
    val nationalNumber: String
) {
    /**
     * Get the full number as digits only (without +).
     */
    val digits: String
        get() = "${countryCode.code}$nationalNumber"

    /**
     * Get the total length of the phone number.
     */
    val length: Int
        get() = digits.length

    /**
     * Format in E.164 format (+CCNNNN...).
     */
    fun formatE164(): String = "+$digits"

    /**
     * Format with spaces for readability (e.g., +1 555 123 4567).
     */
    fun formatInternational(): String {
        val formatted = formatNationalNumber(nationalNumber)
        return "${countryCode.displayPrefix} $formatted"
    }

    override fun toString(): String = formatE164()

    private fun formatNationalNumber(number: String): String {
        val len = number.length
        return when {
            len <= 4 -> number
            len <= 7 -> "${number.substring(0, 3)} ${number.substring(3)}"
            len <= 10 -> "${number.substring(0, 3)} ${number.substring(3, 6)} ${number.substring(6)}"
            else -> number
        }
    }

    companion object {
        /**
         * Parse phone number from string (E.164 format or digits).
         *
         * @param input The phone number string to parse.
         * @return Result containing the parsed PhoneNumber or an error.
         */
        fun parse(input: String): Result<PhoneNumber> = runCatching {
            val trimmed = input.trim()
            if (trimmed.isEmpty()) {
                throw IllegalArgumentException(PhoneError.EmptyInput.toString())
            }

            val (digits, startsWithPlus) = extractDigits(trimmed)
            val digitCount = digits.length

            if (digitCount < 7) {
                throw IllegalArgumentException(PhoneError.TooShort(digitCount).toString())
            }
            if (digitCount > 15) {
                throw IllegalArgumentException(PhoneError.TooLong(digitCount).toString())
            }

            val (countryCode, nationalNumber) = CountryCode.parse(digits)
                ?: throw IllegalArgumentException(
                    PhoneError.InvalidCountryCode(digits.take(3)).toString()
                )

            if (nationalNumber.length < 4) {
                throw IllegalArgumentException(
                    PhoneError.InvalidNationalNumber(nationalNumber).toString()
                )
            }

            PhoneNumber(countryCode, nationalNumber)
        }

        /**
         * Check if string is a valid phone number.
         */
        fun isValid(input: String): Boolean = parse(input).isSuccess

        private fun extractDigits(input: String): Pair<String, Boolean> {
            val startsWithPlus = input.startsWith("+")
            val digits = input.filter { it.isDigit() }
            return Pair(digits, startsWithPlus)
        }
    }
}

/**
 * Safe phone number operations.
 */
object SafePhone {
    /**
     * Parse phone number from string.
     *
     * @param input The phone number string to parse.
     * @return The parsed PhoneNumber or null if invalid.
     */
    fun parse(input: String): PhoneNumber? = PhoneNumber.parse(input).getOrNull()

    /**
     * Parse phone number from string, returning Result.
     *
     * @param input The phone number string to parse.
     * @return Result containing the parsed PhoneNumber or an error.
     */
    fun parseResult(input: String): Result<PhoneNumber> = PhoneNumber.parse(input)

    /**
     * Check if string is a valid phone number.
     *
     * @param input The string to validate.
     * @return True if the string is a valid phone number.
     */
    fun isValid(input: String): Boolean = PhoneNumber.isValid(input)

    /**
     * Format phone number in E.164 format.
     *
     * @param phoneNumber The phone number to format.
     * @return The E.164 formatted string.
     */
    fun formatE164(phoneNumber: PhoneNumber): String = phoneNumber.formatE164()

    /**
     * Format phone number for international display.
     *
     * @param phoneNumber The phone number to format.
     * @return The formatted string with spaces.
     */
    fun formatInternational(phoneNumber: PhoneNumber): String = phoneNumber.formatInternational()

    /**
     * Get the country code from a phone number.
     *
     * @param phoneNumber The phone number.
     * @return The country code.
     */
    fun getCountryCode(phoneNumber: PhoneNumber): CountryCode = phoneNumber.countryCode

    /**
     * Get the national number from a phone number.
     *
     * @param phoneNumber The phone number.
     * @return The national number without country code.
     */
    fun getNationalNumber(phoneNumber: PhoneNumber): String = phoneNumber.nationalNumber

    /**
     * Get the full number as digits only.
     *
     * @param phoneNumber The phone number.
     * @return The digits without + prefix.
     */
    fun getDigits(phoneNumber: PhoneNumber): String = phoneNumber.digits
}
